core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/solver/DefaultSolverJob.java

```diff
@@ -167,13 +167,14 @@ public void addProblemChange(ProblemChange<Solution_> problemChange) {
     public void terminateEarly() {
         try {
             solverStatusModifyingLock.lock();
-            finalBestSolutionFuture.cancel(false);
             switch (solverStatus) {
                 case SOLVING_SCHEDULED:
+                    finalBestSolutionFuture.cancel(false);
```

Comment:
I don't like this much. Is it necessary? Cancel = exceptions.
If solving not yet started, calling terminate should have the same effect, right?

Reply:
I think it is, as `solver.solve()` resets the early termination [1]. Otherwise, we would need to keep a flag and properly synchronize any access to it.

[1] https://github.com/kiegroup/optaplanner/blob/main/core%2Foptaplanner-core-impl%2Fsrc%2Fmain%2Fjava%2Forg%2Foptaplanner%2Fcore%2Fimpl%2Fsolver%2FDefaultSolver.java#L231

Reply:
Agreed to catch the CancellationException and return the input problem in case the solver job is only scheduled.